### PR TITLE
[Fortran/gfortran] Disable tests only on AArch64, but enable them els…

### PR DESCRIPTION
**Pull request by tarunprabhu**
…ewhere

Some tests fail only on AArch64, but pass on other platforms. Previously, these were disabled everywhere, but now they can be disabled only on AArch64 and enabled on other platforms. In addition, some tests that were failing only on AArch64 now pass, so they are enabled everywhere.

The static test configuration has also been updated to reflect these changes. Some bugs were exposed in the update script during this process. These have also been fixed.

----------------------------------------------------------------------------------------
@ylzsx, 

Since these tests were preciously disabled on LoongArch, could you check that they pass there. If they do not, I will update the PR to disable them on LoongArch as well.

### CODE REVIEW

**Comment by ylzsx** [comment]
Thanks, I have tested this pr on a LoongArch machine. Two tests, `findloc_8.f90` and `pr91497.f90`, fail to compile, with the same errors as on aarch64. Please disable these two tests on LoongArch64.
```
--- a/Fortran/gfortran/regression/override.yaml
+++ b/Fortran/gfortran/regression/override.yaml
@@ -25,7 +25,7 @@
 #     Assertion `Elt->getBitWidth() == EltVT.getSizeInBits() && "APInt size does not match type size!"' failed.
 #
 "findloc_8.f90":
-  disabled_on: ["aarch64-*-*"]
+  disabled_on: ["aarch64-*-*", "loongarch64-*-*"]
 
 # entry_23.f raises a segmentation fault at runtime, but only on AArch64.
 "entry_23.f":
@@ -35,7 +35,7 @@
 # error: 'kind=' argument must be a constant scalar integer whose value is a
 # supported kind for the intrinsic result type.
 "pr91497.f90":
-  disabled_on: ["aarch64-*-*"]
+  disabled_on: ["aarch64-*-*", "loongarch64-*-*"]
```

**Comment by jeanPerier** [comment]
+ @kkwli for powerpc

**Review by tblah — APPROVED** [review]
Works for me on aarch64. Thank you Tarun

**Comment by tarunprabhu** [comment]
> Are they issues opened in llvm-project to cover the aarch64 failures?
> 
> If so, please add them in comment here, if not, it would be great to open some, even if the features are not meant to be supported there, it would be best to gracefully fail.

I was planning to start filing issues for some of these once I enabled some more tests, but now is as good a time as any, I suppose.

**Comment by tarunprabhu** [comment]
Issues have been filed for these two tests:

`entry_23.f`: [#119418](https://github.com/llvm/llvm-project/issues/119418)
`findloc_8.f90`: [#119420](https://github.com/llvm/llvm-project/issues/119420)

**Comment by ylzsx** [comment]
> Thanks, I have tested this pr on a LoongArch machine. Two tests, `findloc_8.f90` and `pr91497.f90`, fail to compile, with the same errors as on aarch64. Please disable these two tests on LoongArch64.
> 

@tarunprabhu It seems you have marked the wrong test case; the failing one is `pr91497.f90`, not `entry_23.f` on LoongArch.

**Comment by tarunprabhu** [comment]
> @tarunprabhu It seems you have marked the wrong test case; the failing one is `pr91497.f90`, not `entry_23.f` on LoongArch.

Ackh! Thanks for catching that. Fixed.

**Comment by kkwli** [comment]
I tried the patch on powerpc64le. I got
```
FAIL: gfortran-regression-execute-regression__entry_23_f.test
NOEXE: gfortran-regression-execute-regression__findloc_8_f90.test
PASS: gfortran-regression-execute-regression__pr99210_f90.test
PASS: gfortran-regression-execute-regression__maxlocval_1_f90.test
```
Both `entry_23.f` and `findloc_8.f90` fail with the same behavior as described in https://github.com/llvm/llvm-project/issues/119418 and https://github.com/llvm/llvm-project/issues/119420 respectively. However, when I try to disable these two test cases.
```diff
diff --git a/Fortran/gfortran/regression/override.yaml b/Fortran/gfortran/regression/override.yaml
index 88f312ce..7bed3ead 100644
--- a/Fortran/gfortran/regression/override.yaml
+++ b/Fortran/gfortran/regression/override.yaml
@@ -25,11 +25,11 @@
 #     Assertion `Elt->getBitWidth() == EltVT.getSizeInBits() && "APInt size does not match type size!"' failed.
 #
 "findloc_8.f90":
-  disabled_on: ["aarch64-*-*", "loongarch64-*-*"]
+  disabled_on: ["aarch64-*-*", "loongarch64-*-*", "powerpc64le-*-*"]
 
 # entry_23.f raises a segmentation fault at runtime, but only on AArch64.
 "entry_23.f":
-  disabled_on: ["aarch64-*-*"]
+  disabled_on: ["aarch64-*-*", "powerpc64le-*-*"]
 
 # pr91497.f90 fails to compile on some platforms with the following message:
 # error: 'kind=' argument must be a constant scalar integer whose value is a
diff --git a/Fortran/gfortran/regression/tests.cmake b/Fortran/gfortran/regression/tests.cmake
index 84762ad6..857dcc25 100644
--- a/Fortran/gfortran/regression/tests.cmake
+++ b/Fortran/gfortran/regression/tests.cmake
@@ -4701,7 +4701,7 @@ run;entry_12.f90;;;;
 run;entry_13.f90;;;;
 run;entry_14.f90;;;;
 run;entry_16.f90;;;;
-run;entry_23.f;;;;aarch64-*-* loongarch64-*-*
+run;entry_23.f;;;;aarch64-*-* loongarch64-*-* powerpc64le-*-*
 run;entry_26.f90;;-fno-f2c;;
 run;entry_27.f90;;-ff2c;;
 run;entry_3.f90;;;;
@@ -4806,7 +4806,7 @@ run;findloc_3.f90;;;;
 run;findloc_4.f90;;;;
 run;findloc_5.f90;;;;
 run;findloc_6.f90;;;;
-run;findloc_8.f90;;;;aarch64-*-* loongarch64-*-*
+run;findloc_8.f90;;;;aarch64-*-* loongarch64-*-* powerpc64le-*-*
 run;float_1.f90;;;;
 run;flush_1.f90;;;;
 run;fmt_bz_bn.f;;;;
@@ -6460,4 +6460,4 @@ run;zero_sized_3.f90;;;;
 run;zero_sized_4.f90;;;;
 run;zero_sized_5.f90;;;;
 run;zero_sized_8.f90;;;;
-run;zero_sized_9.f90;;;;
\ No newline at end of file
+run;zero_sized_9.f90;;;;
```
I got the same outcome.
```
FAIL: gfortran-regression-execute-regression__entry_23_f.test
NOEXE: gfortran-regression-execute-regression__findloc_8_f90.test
PASS: gfortran-regression-execute-regression__pr99210_f90.test
PASS: gfortran-regression-execute-regression__maxlocval_1_f90.test
```

**Comment by tarunprabhu** [comment]
Thanks for checking, Kelvin. It's still odd that you are getting the NOEXE. That suggests that it is attempting to build `findloc_8.f90` and failing. Do you mind checking what ${CMAKE_SYSTEM_PROCESSOR} and ${CMAKE_SYSTEM_NAME} return on your system? So far, that is the best hypothesis. We should try to get this resolved now.

**Comment by kkwli** [comment]
> Thanks for checking, Kelvin. It's still odd that you are getting the NOEXE. That suggests that it is attempting to build `findloc_8.f90` and failing. Do you mind checking what ${CMAKE_SYSTEM_PROCESSOR} and ${CMAKE_SYSTEM_NAME} return on your system? So far, that is the best hypothesis. We should try to get this resolved now.

```
$ cat CMakeLists.txt 
message(STATUS "CMAKE_SYSTEM_PROCESSOR='${CMAKE_SYSTEM_PROCESSOR}'")
message(STATUS "CMAKE_SYSTEM_NAME='${CMAKE_SYSTEM_NAME}'")
```

Output:
```
...
-- CMAKE_SYSTEM_PROCESSOR='ppc64le'
-- CMAKE_SYSTEM_NAME='Linux'
...
```

**Comment by ylzsx** [comment]
@tarunprabhu I sincerely apologize. After re-verifying, I found that `entry_23.f` also produces a similar error on LoongArch as it does on AArch64. Previously, running `ninja check` showed that the test passed, so I didn’t pay much attention. However, when I ran the test independently today, it also resulted in a segmentation fault.

Digging deeper into the cause, I discovered that during the CMake configuration, I had set the parameter `TEST_SUITE_USE_PERF=ON`, and with this setting, running `ninja check` passed without any issues. However, when `TEST_SUITE_USE_PERF=OFF` is set, the error occurs. `TEST_SUITE_USE_PERF` is defined https://github.com/llvm/llvm-test-suite/blob/main/tools/CMakeLists.txt#L23

I’m not sure whether it’s an issue with `perf` or `timeit.sh`. However, `entry_23.f` does indeed fail on LoongArch. I kindly ask you to mark it again. Once again, I sincerely apologize for my previous misunderstanding.

To summarize, it has now been found that a total of three test cases fail on LoongArch: `findloc_8.f90` , `pr91497.f90` and `entry_23.f`.

**Comment by tarunprabhu** [comment]
That's quite alright, Zhaoxin. Happens to all of us :-)

It's curious that the test is reported as a pass with `timeit.sh`. I might take a look at it, but not right now.

**Comment by tarunprabhu** [comment]
Thanks for checking, Kelvin. I'll take a closer look to try and understand what might be happening.

**Comment by tarunprabhu** [comment]
I think one of the issues might have been that `*` is allowed as a wildcard match in the override files. But I don't think it can be used that way in cmake regexes. When parsing the DejaGNU directives from the test files, those are converted to `.+` which is what cmake expects. But this was not done when parsing the override files. The latest commit fixes this. 

@kkwli, could you check if the tests can be disabled on PowerPC with this change?

**Review by ylzsx — APPROVED** [review]

**Comment by kkwli** [comment]
It works with using `ppc64le` (not `powerpc64le`). Thanks a lot.

```diff
diff --git a/Fortran/gfortran/regression/override.yaml b/Fortran/gfortran/regression/override.yaml
index 7b3e2e81..d7e5fa17 100644
--- a/Fortran/gfortran/regression/override.yaml
+++ b/Fortran/gfortran/regression/override.yaml
@@ -25,11 +25,11 @@
 #     Assertion `Elt->getBitWidth() == EltVT.getSizeInBits() && "APInt size does not match type size!"' failed.
 #
 "findloc_8.f90":
-  disabled_on: ["aarch64-*-*", "loongarch64-*-*"]
+  disabled_on: ["aarch64-*-*", "loongarch64-*-*", "ppc64le-*-*"]
 
 # entry_23.f raises a segmentation fault at runtime, on some platforms.
 "entry_23.f":
-  disabled_on: ["aarch64-*-*", "loongarch64-*-*"]
+  disabled_on: ["aarch64-*-*", "loongarch64-*-*", "ppc64le-*-*"]
 
 # pr91497.f90 fails to compile on some platforms with the following message:
 # error: 'kind=' argument must be a constant scalar integer whose value is a
diff --git a/Fortran/gfortran/regression/tests.cmake b/Fortran/gfortran/regression/tests.cmake
index 549182d5..e5cc6bf3 100644
--- a/Fortran/gfortran/regression/tests.cmake
+++ b/Fortran/gfortran/regression/tests.cmake
@@ -4701,7 +4701,7 @@ run;entry_12.f90;;;;
 run;entry_13.f90;;;;
 run;entry_14.f90;;;;
 run;entry_16.f90;;;;
-run;entry_23.f;;;;aarch64-.+-.+ loongarch64-.+-.+
+run;entry_23.f;;;;aarch64-.+-.+ loongarch64-.+-.+ ppc64le-.+-.+
 run;entry_26.f90;;-fno-f2c;;
 run;entry_27.f90;;-ff2c;;
 run;entry_3.f90;;;;
@@ -4806,7 +4806,7 @@ run;findloc_3.f90;;;;
 run;findloc_4.f90;;;;
 run;findloc_5.f90;;;;
 run;findloc_6.f90;;;;
-run;findloc_8.f90;;;;aarch64-.+-.+ loongarch64-.+-.+
+run;findloc_8.f90;;;;aarch64-.+-.+ loongarch64-.+-.+ ppc64le-.+-.+
 run;float_1.f90;;;;
 run;flush_1.f90;;;;
 run;fmt_bz_bn.f;;;;
```

**Review by kkwli — APPROVED** [review]
LG. Thanks.
